pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go

```diff
@@ -58,161 +57,246 @@ type rangefeedMuxer struct {
+			if err != nil {
+				return err
+			}
+
+			if err := conn.startRangeFeed(streamID, stream, &args); err != nil {
+				log.VErrEventf(ctx, 1,
+					"RPC error establishing mux rangefeed to replica %s: %s", args.Replica, err)
+				continue
+			}
+			// We successfully established rangefeed, so the responsibility
+			// for releasing the stream is transferred to the mux go routine.
+			stream = nil
+			return nil
 		}
-	}
 
-	if err := ms.client.Send(req); err != nil {
-		cleanup()
-		return nil, nil, err
+		// If the transport is exhausted, we evict routing token and retry range
+		// resolution.
+		token.Evict(ctx)
+		token = rangecache.EvictionToken{}
 	}
-	return producer, cleanup, nil
+
+	return ctx.Err()
 }
 
 // establishMuxConnection establishes MuxRangeFeed RPC with the node.
 func (m *rangefeedMuxer) establishMuxConnection(
 	ctx context.Context, client rpc.RestrictedInternalClient, nodeID roachpb.NodeID,
-) (*muxClientState, error) {
-	// NB: the `ctx` in scope here belongs to a client for a single range feed, and must
-	// not influence the lifetime of the mux connection. At the time of writing, the caller
-	// is `singleRangeFeed` which calls into this method through its streamProducerFactory
-	// argument.
-	m.mu.Lock()
-	ms, found := m.mu.clients[nodeID]
-	if !found {
-		// Initialize muxClientState.
-		// Only initCtx is initialized here since we need to block on it.
-		// The rest of the initialization happens in startNodeMuxRangeFeed.
-		ms = &muxClientState{initCtx: makeTerminationContext()}
-		// Kick off client initialization on another Go routine.
-		// It is important that we start MuxRangeFeed RPC using long-lived
-		// context available in the main context group used for this muxer.
-		m.g.GoCtx(func(ctx context.Context) error {
-			return m.startNodeMuxRangeFeed(ctx, ms, client, nodeID)
-		})
-		m.mu.clients[nodeID] = ms
+) (*muxNodeConn, error) {
+	var muxClient *future.Future[connOrError]
+
+	{
+		ptr, exists := m.muxClients.Load(int64(nodeID))
+		if !exists {
+			ptr, exists = m.muxClients.LoadOrStore(int64(nodeID), unsafe.Pointer(future.MakePromise[connOrError]()))
+		}
+		muxClient = (*future.Future[connOrError])(ptr)
+		if !exists {
+			// Kick off client initialization on another Go routine.
+			m.g.GoCtx(func(ctx context.Context) error {
+				return m.startNodeMuxRangeFeed(ctx, client, nodeID, muxClient)
+			})
+		}
 	}
-	ms.numStreams++
-	m.mu.Unlock()
 
 	// Ensure mux client is ready.
+	init := future.MakeAwaitableFuture(muxClient)
 	select {
 	case <-ctx.Done():
 		return nil, ctx.Err()
-	case <-ms.initCtx.Done():
-		return ms, ms.initCtx.Err()
+	case <-init.Done():
+		c := init.Get()
+		return c.conn, c.err
 	}
 }
 
 // startNodeMuxRangeFeedLocked establishes MuxRangeFeed RPC with the node.
 func (m *rangefeedMuxer) startNodeMuxRangeFeed(
 	ctx context.Context,
-	ms *muxClientState,
 	client rpc.RestrictedInternalClient,
 	nodeID roachpb.NodeID,
-) error {
+	conn *future.Future[connOrError],
```

Comment:
I'm finding it really confusing that we're passing in both `client` and `conn` here. Consider lifting all of the stream setup stuff above here, and make it clear that `conn` is not an input per sé but rather a result that will be populated from `client`. Maybe also consider using `stream` rather than `conn` for each individual multiplexed stream, since `conn` reads more like the underlying gRPC transport connection rather than the logical multiplexed stream (which logically wouldn't be possible since `client` depends on the transport connection).